libnavigation-core/src/main/java/com/mapbox/navigation/core/stops/ArrivalProgressObserver.kt

```diff
@@ -0,0 +1,71 @@
+package com.mapbox.navigation.core.stops
+
+import com.mapbox.navigation.base.trip.RouteProgressObserver
+import com.mapbox.navigation.base.trip.model.RouteLegProgress
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.trip.session.TripSession
+
+internal class ArrivalProgressObserver(
+    private val tripSession: TripSession
+) : RouteProgressObserver {
+
+    private var arrivalController: ArrivalController = AutoArrivalController()
+
+    fun attach(arrivalController: ArrivalController) {
+        this.arrivalController = arrivalController
+        tripSession.getRouteProgress()?.let {
+            onRouteProgressChanged(it)
+        }
+    }
+
+    fun detach() {
+        arrivalController = AutoArrivalController()
+    }
+
+    fun navigateToNextStop(): Boolean {
+        val routeIndex = tripSession.getRouteProgress()?.route()?.routeIndex()?.toInt()
+            ?: return false
+        val numberOfLegs = tripSession.getRouteProgress()?.route()?.legs()?.size
+            ?: return false
+        val legIndex = tripSession.getRouteProgress()?.currentLegProgress()?.legIndex()
+            ?: return false
+        val nextLegIndex = legIndex + 1
+        return if (nextLegIndex < numberOfLegs) {
+            tripSession.updateRouteLegIndex(routeIndex, nextLegIndex)
+            false
+        } else {
+            true
+        }
+    }
+
+    override fun onRouteProgressChanged(routeProgress: RouteProgress) {
+        val routeLegProgress = routeProgress.currentLegProgress()
+            ?: return
+
+        val arrivalOptions = arrivalController.arrivalOptions()
+        if (arrivalOptions.arrivalInSeconds != null) {
+            checkArrivalTime(arrivalOptions.arrivalInSeconds, routeLegProgress)
+        } else if (arrivalOptions.arrivalInMeters != null) {
+            checkArrivalDistance(arrivalOptions.arrivalInMeters, routeLegProgress)
+        }
+    }
+
+    private fun checkArrivalTime(arrivalInSeconds: Double, routeLegProgress: RouteLegProgress) {
+        if (routeLegProgress.durationRemaining() <= arrivalInSeconds) {
+            doOnArrival(routeLegProgress)
+        }
+    }
+
+    private fun checkArrivalDistance(arrivalInMeters: Double, routeLegProgress: RouteLegProgress) {
+        if (routeLegProgress.distanceRemaining() <= arrivalInMeters) {
+            doOnArrival(routeLegProgress)
+        }
+    }
+
+    private fun doOnArrival(routeLegProgress: RouteLegProgress) {
+        val moveToNextLeg = arrivalController.onArrival(routeLegProgress)
+        if (moveToNextLeg) {
+            navigateToNextStop()
+        }
```

Comment:
Should we expect the navigation is paused when `onArrival` return false?

Reply:
would expect navigation to continue normally, return false is a no-op. intention is to communicate, i have not arrived.

return true should proceed to next navigation, arrival has been completed